Реализация функций TO-DO приложения

subbbbbaru opened the pull request:
Реализованы следующие функции:
getAllTasks
getTaskById
addTask
deleteTask

subbbbbaru committed:
init commit
Реализованы следующие функции:
getAllTasks
getTaskById
addTask
deleteTask

subbbbbaru committed:
fix write error in addTask

subbbbbaru committed:
test autotest

ivannizh commented on precode.go:
```diff
@@ -40,13 +43,79 @@ var tasks = map[string]Task{
+func getAllTasks(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	w.WriteHeader(http.StatusOK)
```
Это пишется в конце, так как если ошибка, то хедер со статусом записан 200

ivannizh commented on precode.go:
```diff
@@ -40,13 +43,79 @@ var tasks = map[string]Task{
 // Ниже напишите обработчики для каждого эндпоинта
-// ...
+func getAllTasks(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
```
Это в конец нужно ставить перед записью ответа, так как ошибка, например, отдается в другом формате

ivannizh commented on precode.go:
```diff
@@ -40,13 +43,79 @@ var tasks = map[string]Task{
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
```
Аналогично

ivannizh commented on precode.go:
```diff
@@ -40,13 +43,79 @@ var tasks = map[string]Task{
+	err := json.NewEncoder(w).Encode(task)
+	if err != nil {
+		log.Printf("ошибка кодирования JSON: %s", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Так как это внутренняя ошибка, то лучше использовать InternalServerError

ivannizh commented on precode.go:
```diff
@@ -40,13 +43,79 @@ var tasks = map[string]Task{
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
```
Аналогично

ivannizh replied:
Если нет тела ответа, то можно и не выставлять тип контента.

ivannizh commented on precode.go:
```diff
@@ -40,13 +43,79 @@ var tasks = map[string]Task{
+		http.Error(w, "Id not found", http.StatusBadRequest)
+		return
+	}
+	tasks[newTask.ID] = newTask
```
При таком подходе если задача с таким айди уже было, то она будет перезатерта. По ТЗ это ок, но лучше помнить об этом

ivannizh reviewed (commented):
Привет! Хорошая работа, но нужно немного подправить

subbbbbaru committed:
add .gitignore
fix body in func(getAllTasks, getTaskById, addTask) and write correct return value

ivannizh commented on precode.go:
```diff
@@ -44,51 +44,69 @@ var tasks = map[string]Task{
+	_, err = w.Write(resp)
+	if err != nil {
+		log.Printf("ошибка отправки ответа JSON: %s", err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Не отписал в прошлый раз, что если ошибка записи ответа, то нет смысла в тот же ответ писать ошибку, потому что это все равно не получится сделать. Такие ошибка просто логируют, чтобы потом уже смотреть причину сбоя.

Такие ошибки могут быть при отмене запроса и уже точно не получится туда записать ошибку

subbbbbaru replied:
Спасибо большое! Учту в будущем.

ivannizh commented on precode.go:
```diff
@@ -44,51 +44,69 @@ var tasks = map[string]Task{
+	resp, err := json.Marshal(task)
+	if err != nil {
+		log.Printf("ошибка кодирования JSON: %s", err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Нет выхода из функции, сначала запишем ошибку, а потом хороший ответ

ivannizh reviewed (approved):
Привет! Правки приняты!

subbbbbaru committed:
fix code in bad response writer
